src/main/java/io/r2dbc/proxy/listener/LifeCycleListener.java

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.r2dbc.proxy.listener;
+
+import io.r2dbc.proxy.core.MethodExecutionInfo;
+import io.r2dbc.proxy.core.QueryExecutionInfo;
+
+/**
+ * Provides callback for each SPI.
+ *
+ * @author Tadaya Tsuyukubo
+ * @see LifeCycleExecutionListener
+ */
+public interface LifeCycleListener {
+
+    //
+    // for ConnectionFactory
+    //
+
+    default void beforeCreateOnConnectionFactory(MethodExecutionInfo methodExecutionInfo) {
```

Comment:
Need documentation on these methods as well, _especially_ around nullability guarantees for implementors.